chemprop/nn/metrics.py

```diff
@@ -63,160 +77,473 @@ def forward(
+    """
+
+    def __init__(self, task_weights: ArrayLike = 1.0):
+        """
+        Parameters
+        ----------
+        task_weights :  ArrayLike, default=1.0
+            the per-task weights of shape `t` or `1 x t`. Defaults to all tasks having a weight of 1.
+        """
+        super().__init__()
+        task_weights = torch.as_tensor(task_weights, dtype=torch.float).view(1, -1)
+        self.register_buffer("task_weights", task_weights)
+
+        self.add_state("p", default=[], dist_reduce_fx="cat")
+        self.add_state("t", default=[], dist_reduce_fx="cat")
+        self.add_state("c", default=torch.tensor(0.0), dist_reduce_fx="sum")
+        self.add_state("s", default=torch.tensor(0.0), dist_reduce_fx="sum")
+
+    def update(
+        self,
+        preds: Tensor,
+        targets: Tensor,
+        mask: Tensor | None = None,
+        weights: Tensor | None = None,
+        *args,
+    ):
+        mask = torch.ones_like(targets, dtype=torch.bool) if mask is None else mask
+        weights = torch.ones_like(targets, dtype=torch.float) if weights is None else weights
+
+        if not (0 <= preds.min() and preds.max() <= 1):  # assume logits
+            preds = preds.softmax(2)
+
+        p, t, c, s = self._calc_unreduced_loss(preds, targets.long(), mask, weights, *args)
+
+        self.p += [p]
+        self.t += [t]
+        self.c += c
+        self.s += s
+
+    def _calc_unreduced_loss(self, preds, targets, mask, weights, *args) -> Tensor:
+        device = preds.device
+        C = preds.shape[2]
+        bin_targets = torch.eye(C, device=device)[targets]
+        bin_preds = torch.eye(C, device=device)[preds.argmax(-1)]
+        masked_data_weights = weights.unsqueeze(2) * mask.unsqueeze(2)
+        p = (bin_preds * masked_data_weights).sum(0)
+        t = (bin_targets * masked_data_weights).sum(0)
+        c = (bin_preds * bin_targets * masked_data_weights).sum()
+        s = (preds * masked_data_weights).sum()
+
+        return p, t, c, s
+
+    def compute(self):
+        p = torch.stack(self.p, dim=-1).sum(-1)
+        t = torch.stack(self.t, dim=-1).sum(-1)
+        c = self.c
+        s = self.s
+        s2 = s.square()
+
+        # the `einsum` calls amount to calculating the batched dot product
+        cov_ytyp = c * s - torch.einsum("ij,ij->i", p, t).sum()
+        cov_ypyp = s2 - torch.einsum("ij,ij->i", p, p).sum()
+        cov_ytyt = s2 - torch.einsum("ij,ij->i", t, t).sum()
+
+        x = cov_ypyp * cov_ytyt
+        MCC = torch.tensor(0.0, device=x.device) if x == 0 else cov_ytyp / x.sqrt()
+        MCC = MCC * self.task_weights
+
+        return 1 - MCC.mean()
 
 
 @MetricRegistry.register("multiclass-mcc")
-class MulticlassMCCMetric(MulticlassMCCLoss, Metric):
+class MulticlassMCCMetric(MulticlassMCCLoss):
+    def compute(self):
+        return 1 - super().compute()
+
+
+class ChempropClassificationMixin:
```

Comment:
```suggestion
class ClassificationMixin:
```

Reply:
Also, all the `Mixin` functions are missed in `__init__.py`

Reply:
I took inspiration from classes like `_MolGraphDatasetMixin`. They are no supposed to be imported by users so they are not included in the init.py file, but can still be imported directly when needed. (`from chemprop.data.datasets import _MolGraphDatasetMixin`). I will add a `_` to the name of the mixin to suggest that.